### PR TITLE
Miners' hashrate just do it version

### PR DESCRIPTION
**Pull request by sgaragagghu**
I need time to study it in depth, so meanwhile i push it, that works good.
Please give me feedback.
The mean is around 4h.

### CODE REVIEW

File: client/pool.c [hunk 1]
```diff
@@ -1419,3 +1427,27 @@ void* pool_remove_inactive_connections(void* arg)
 	return NULL;
 }
 
+static inline void update_mean_log_diff(struct connection_pool_data *conn_data, struct xdag_pool_task *task, xdag_hash_t hash)
```
Comment (jonano614) [inline client/pool.c]: It is too large function to be inline.

File: client/pool.c [hunk 2]
```diff
@@ -1419,3 +1427,27 @@ void* pool_remove_inactive_connections(void* arg)
...
+
+	if(conn_data->task_time < task_time) {
+		moving_average(&conn_data->mean_log_difficulty, diff2log(hash2difficulty(conn_data->last_min_hash)),conn_data->bounded_task_counter); 
+                if(conn_data->bounded_task_counter<NSAMPLES_MAX)
+			++conn_data->bounded_task_counter;
+		memcpy(conn_data->last_min_hash,hash,sizeof(xdag_hash_t));
+	} else if(xdag_cmphash(hash, conn_data->last_min_hash) < 0)
+		memcpy(conn_data->last_min_hash,hash,sizeof(xdag_hash_t));
+
+	if(conn_data->miner) {
```
Comment (jonano614) [inline client/pool.c]: conn_data->miner cannot be NULL here.

File: client/utils/moving_statistics/moving_average.c [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+If you modify this file, you have to release the source code of this file included all of the modifications apported.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+***************************************************************************/
+
+#include "moving_average.h"
+
+
+void welford_one_pass(long double* mean, long double sample, uint16_t nsamples){
```
Comment (jonano614) [inline client/utils/moving_statistics/moving_average.c]: Can you make the name of function more descriptive? welford_one_pass - it is a bit unclear what this function does.

File: client/utils/moving_statistics/moving_average.c [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	if(nsamples)
+        	*mean=*mean+(sample-*mean)/(long double)(nsamples);
+}
+
+
+void moving_average(long double* mean, long double sample, uint16_t nsamples){
```
Comment (jonano614) [inline client/utils/moving_statistics/moving_average.c]: Add a comment please what this function does.

File: client/utils/moving_statistics/moving_average.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#ifndef _MOVING_AVERAGE_H
+#define _MOVING_AVERAGE_H
+
+#include <stdint.h>
+
+#ifndef NSAMPLES_MAX
+#define NSAMPLES_MAX 255 
+#endif
+
+void welford_one_pass(long double* mean, long double sample, uint16_t nsamples);
```
Comment (jonano614) [inline client/utils/moving_statistics/moving_average.h]: You use that function only in moving_average.c. You don't need to declare it in the header file.

File: client/utils/utils.h [hunk 6]
```diff
@@ -12,6 +11,7 @@
 #include <stdio.h>
 #include <stdint.h>
 #include <pthread.h>
+#include "math.h"
```
Comment (jonano614) [inline client/utils/utils.h]: Do you really this header here?

File: client/pool.c [hunk 7]
```diff
@@ -95,6 +100,9 @@ struct connection_pool_data {
 	time_t last_share_time;
 	int deleted;
 	char* disconnection_reason;
+        xdag_hash_t last_min_hash;
```
Comment (jonano614) [inline client/pool.c]: Use tabs

File: client/pool.c (hunk 2 above)
Comment (sgaragagghu) [reply client/pool.c]: You did it this way in your function too

File: client/block.c [hunk 8]
```diff
@@ -320,8 +320,10 @@ static inline xdag_diff_t hash_difficulty(xdag_hash_t hash)
 	xdag_diff_t res = ((xdag_diff_t*)hash)[1], max = xdag_diff_max;
 
 	xdag_diff_shr32(&res);
-	
-	return xdag_diff_div(max, res);
+
+	if(res)
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: not NULL, but 0. But it shouldn't never happen to be 0. Try to do a 0 division there, it's dangerous.

File: client/utils/moving_statistics/moving_average.c (hunk 3 above)
Comment (sgaragagghu) [reply client/utils/moving_statistics/moving_average.c]: It's a famous algorithm, i'll add a small comment to explain

File: client/block.c [hunk 9]
```diff
@@ -1486,3 +1488,9 @@ int xdagGetLastMainBlocks(int count, char** addressArray)
 	}
 	return i;
 }
+
+xdag_diff_t hash2difficulty(xdag_hash_t hash)
```
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: To keep the original function inline, the original function is called continuously in block.c, so i left it inline for performance (even if i should check assembly to understand if the compiler really does the inline)

File: client/pool.c (hunk 2 above)
Comment (sgaragagghu) [reply client/pool.c]: you do it in calculate_nopaid_shares

File: client/block.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (sgaragagghu) [reply client/block.c]: xdag_diff is a 128 bit uint on linux
a struct of 4*32bit uint  in windows

**Comment by sgaragagghu** [comment]
fixed conflicts

**Comment by sgaragagghu** [comment]
refixed conflicts

**Comment by jonano614** [comment]
Thank you for your great efforts!

**Pull request merged**